84-slurm (#103)
ci-for-research/self-hosted-runners: ubuntu-docker/README.md

# Setting up a CI server for a GitHub Action runner with Docker from Linux Ubuntu

> After following this guide, you'll have a simple GitHub action workflow on a GitHub repository of your choice. When new commits are made to your repository, the workflow delegates work to a server which runs in a [Docker](https://www.docker.com/) container. You can follow these instructions on your own computer or a Linux server.

Reviewer: Text is clearer without the last part
```suggestion
After following this guide, you'll have a simple GitHub action workflow on a GitHub repository of your choice. When new commits are made to your repository, the workflow delegates work to a server which runs in a [Docker](https://www.docker.com/) container.
```

This guide distinguishes between the _client_ and the _server_; the client is your own machine; the server is whichever
machine will run the tests. This document describes the case where the server is a Docker container running on your own machine.

For guides on how to configure other features in addition to just the runner, go [here](/README.md).

## Prerequisites

1. Install Docker: https://docs.docker.com/engine/install/

> 2. Follow post-installation steps https://docs.docker.com/engine/install/linux-postinstall/ to manage docker as a non-root user

Reviewer: ```suggestion
2. Follow post-installation steps https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user to manage docker as a non-root user
```

> 2. Follow post-installation steps https://docs.docker.com/engine/install/linux-postinstall/

Reviewer: Be more specific about what users need to do. I'm guessing you meant:
```suggestion
2. Follow post-installation steps https://docs.docker.com/engine/install/linux-postinstall/ to manage docker as a non-root user
```

### Testing your Docker setup

Refence: https://docs.docker.com/docker-for-windows/#test-your-installation

> 1. Open a terminal window (Command Prompt or PowerShell, but not PowerShell ISE).

Reviewer: We don't have Command Prompt, PowerShell, PowerShell ISE since we're on Ubuntu Linux
```suggestion
1. Open a terminal window
```

> 2. Run docker --version to ensure that you have a supported version of Docker:

Reviewer: ```suggestion
2. Run ``docker --version`` to ensure that you have a supported version of Docker:
```

Reviewer: maybe specify a minimum recommended version, for example mine is 18.09, should i update?

```

3. Pull the hello-world image from Docker Hub and run a container:

```shell
> docker run hello-world

Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
0e03bdcc26d7: Pull complete
Digest: sha256:d58e752213a51785838f9eed2b7a498ffa1cb3aa7f946dda11af39286c3db9a9
Status: Downloaded newer image for hello-world:latest

Hello from Docker!
This message shows that your installation appears to be working correctly.
...
```

4. List the hello-world image that was downloaded from Docker Hub:

```shell
> docker image ls

REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
hello-world         latest              bf756fb1ae65        6 months ago        13.3kB
```

5. List the hello-world container (that exited after displaying “Hello from Docker!”):

```shell
> docker container ls --all

CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS                          PORTS               NAMES
1d624a063f22        hello-world         "/hello"            About a minute ago   Exited (0) About a minute ago                       flamboyant_ramanujan
```

## Server side configuration

> E.g. how to configure VirtualBox, how to run docker container, how to configure HPC cloud machine
> 

Reviewer: ```suggestion
```

### Build image

> Now we are ready to build our Docker image. The following command will use [Dockerfile](docker/Dockerfile) in `docker` folder to build the image. It will create a system user, install necassary system packages and dependencies for the runner.

Reviewer: ```suggestion
Now we are ready to build our Docker image. The following command will use [Dockerfile](docker/Dockerfile) in `docker` folder to build the image. It will create a system user, install necessary system packages and dependencies for the runner.
```

> Now we are ready to build our Docker image. The following command will use [Dockerfile](docker/Dockerfile) in `docker` folder to build the image. It will create a system user, install necassary system packages and dependencies for the runner.

Reviewer: I'd prefer ``Dockerfile`` and ``entrypoint.sh`` to be at the same level as the ``README.md``

```shell
docker build \

>     -t ga-runner \

Reviewer: I prefer full arg names when explaining things
```suggestion
    --tag ga-runner \
```

    --build-arg DOCKER_USER="<username>" \
    --build-arg DOCKER_PASS="<user password>" \
    ./docker
```

> You will need to adjust `<username>` and `<user password>` for the normal user which will be added to the Docker image.

Reviewer: Are their downsides to being specific here? So instead of the generic ``<username>`` and ``<password>``, why not use ``tester`` and ``password``?

> You will need to adjust `<username>` and `<user password>` for the normal user which will be added to the Docker image.

Reviewer: ```suggestion
You will need to adjust `<username>` and `<user password>` for the user which will be added to the Docker image.
```

> You will need to adjust `<username>` and `<user password>` for the normal user which will be added to the Docker image.

Reviewer: ```suggestion
You will need to adjust `<username>` and `<user password>` for the normal user that will be added to the Docker image.
```

## Client side configuration

### Run the server

#### Temporary mode

> The command below will run the docker image and setup the runner. When user presses `CTRL+C`, it automatically removes the runner from GitHub and removes the Docker container as well.

Reviewer: setup is a noun, set up is a verb

```suggestion
The command below will run the docker image and set up the runner. When user presses `CTRL+C`, it automatically removes the runner from GitHub and removes the Docker container as well.
```
